packages/playground/CoreApp/App.cpp

```diff
@@ -0,0 +1,75 @@
+#include "pch.h"
+#include "CoreApp.h"
+
+// This app demonstrates the usage of the Microsoft.ReactNative CoreApp APIs
+// use the preprocessor defines WITH_MODULES (true/false), FROM_JSON (true/false),
+// or neither to exercise the RNStartCoreAppWithModules, RNStartCoreApp, and RNStartCoreAppFromConfigJson APIs.
+#define DEFINE_MODULES
+
+// #define WITH_MODULES
+#define FROM_JSON
+
+#ifdef DEFINE_MODULES
+#include <ModuleRegistration.h>
+#include <NativeModules.h>
+#include <winrt/Microsoft.ReactNative.h>
+
+REACT_MODULE(LogBox)
+struct LogBox {
+  REACT_INIT(Initialize)
+  void Initialize(const winrt::Microsoft::ReactNative::ReactContext &) noexcept {
+    MessageBox(nullptr, L"Native Modules work!", L"Success", MB_OK);
+  }
+};
+
+struct ThisAppPackageProvider
+    : winrt::implements<ThisAppPackageProvider, winrt::Microsoft::ReactNative::IReactPackageProvider> {
+  void CreatePackage(winrt::Microsoft::ReactNative::IReactPackageBuilder const &packageBuilder) noexcept {
+    for (auto const *reg = winrt::Microsoft::ReactNative::ModuleRegistration::Head(); reg != nullptr;
+         reg = reg->Next()) {
+      packageBuilder.AddModule(reg->ModuleName(), reg->MakeModuleProvider());
+    }
+  }
+};
+
+extern "C" __declspec(dllexport) void *__cdecl MySpecialPackageProvider() {
+  auto provider = winrt::make<ThisAppPackageProvider>();
+  void *abi{nullptr};
+  winrt::copy_to_abi(provider, abi);
+  return abi;
+}
+
+#endif
+
+#ifdef FROM_JSON
+int __stdcall wWinMain(_In_ HINSTANCE, _In_opt_ HINSTANCE, _In_ PWSTR /*server*/, _In_ int /*showCommand*/) {
+  RNStartCoreAppFromConfigJson(L"app.config.json", nullptr);
+}
+#else
+int __stdcall wWinMain(_In_ HINSTANCE, _In_opt_ HINSTANCE, _In_ PWSTR /*server*/, _In_ int /*showCommand*/) {
+
+#ifdef WITH_MODULES
+  RNStartCoreAppWithModules(
+      [](RNCoreApp *app) {
```

Comment:
>[] [](http://example.com/codeflow?start=6&length=2)

The current API does not allow passing any "user" data inside of the lambda.
The better way is to add an additional "void* data" parameter to the RNStartCoreApp function which is then passed to the function:
[](RNCoreApp *app, void* data) { ... }
This way the app can pass a pointer to one of its object if needed.
Better yet, we can define a C++ template function that internally wraps RNStartCoreApp  and allows passing any arbitrary lambda with captured variables.
```
template <typename TLambda>
void RNStartCoreAppWrapper(const TLambda& onStart) {
   RNStartCoreApp([](RNCoreApp *app, void* data) { 
  auto myOnStart = reinterpret_cast<const TLambda*>(data);
   (*myOnStart)(app);
}, &onStart);
}
```

Reply:
addressing the void* param; we can tackle the template stuff later if needed :)